pkg/api/v20210131preview/validate.go

```diff
@@ -0,0 +1,23 @@
+package v20210131preview
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"regexp"
+)
+
+// Regular expressions used to validate the format of resource names and IDs acceptable by API.
+var (
+	RxClusterID       = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]/providers/Microsoft\.RedHatOpenShift/openShiftClusters/[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+	RxResourceGroupID = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+	RxSubnetID        = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]/providers/Microsoft\.Network/virtualNetworks/[-a-z0-9_.]{2,64}/subnets/[-a-z0-9_.]{2,80}$`)
+	RxDomainName      = regexp.MustCompile(`^` +
+		`([a-z][-a-z0-9]{0,61}[a-z0-9])` +
+		`(\.([a-z0-9]|[a-z0-9][-a-z0-9]{0,61}[a-z0-9]))*` +
+		`$`)
+	RxDomainNameRFC1123 = regexp.MustCompile(`^` +
```

Comment:
Added to get rid of import cyclic error.  Also validation from own API is necessary because new VM types are not apart of `pkg/api`

Reply:
Overall I think we need to break this somehow differently. Reasoning being that now it moves into api, but we use this api validation everywhere (operator and dynamic). So this looks like layering violation. 

If we have circular imports maybe these should move somewhere totally outside any api as we gonna use them in both APIS either way. 

So VM types stays API bound, but we split validation outside. Or we even could experiment with something like this:
```
apiV1VMS = []string{
api.VMSizeStandardD4asV4,
api.VMSizeStandardD8asV4,
api.VMSizeStandardD16asV4,
api.VMSizeStandardD32asV4,
api.VMSizeStandardD4sV3
}

apiV2VMS = []{
apiV1VMS,
api.VMSizeStandardD8sV3,
api.VMSizeStandardD16sV3,
api.VMSizeStandardD32sV3,
api.VMSizeStandardE4sV3,
api.VMSizeStandardE8sV3,
}
```
Where we make sure every API version exports VM list for that version, and we already know each api version will be linear support. So we just append previous api into Validator, so making validators API specific...
Just an idea to snowball around.

Reply:
If we are adding new machine types to the newer versions of the API, the only validation that would be different would be quota.  Pinning that to a specific version is fine, but all the other logic will stay the same. 

Using the above, validating the quota usage will be different across the versions but the latest version will be able to validate quota on any version.  Then importing this to allow quota to be used (right now it's tied to the method for dynamic validator) as a standalone function instead of a method on the validator type. So the only new function you would write on future versions of the newly added VM types would be the validatequota separate from what we already do.